File: src/main/java/com/example/demo/util/IpUtils.java
package com.example.demo.util;

import jakarta.servlet.http.HttpServletRequest;
import lombok.experimental.UtilityClass;

@UtilityClass
public class IpUtils {

    public String clientIp(HttpServletRequest request, boolean trustProxy) {
        if (trustProxy) {
            String fwd = request.getHeader("X-Forwarded-For");
            if (fwd != null && !fwd.isBlank()) {
                String first = fwd.split(",")[0].trim();
                if (!first.isBlank()) {
                    return first;
                }
            }
        }
        return request.getRemoteAddr();
    }
}

File: src/main/kotlin/com/example/demo/util/IpUtils.kt
package com.example.demo.util

import jakarta.servlet.http.HttpServletRequest

object IpUtils {
    fun clientIp(request: HttpServletRequest, trustProxy: Boolean): String {
        if (trustProxy) {
            val fwd = request.getHeader("X-Forwarded-For")
            if (!fwd.isNullOrBlank()) {
                val first = fwd.split(',').firstOrNull()?.trim()
                if (!first.isNullOrBlank()) return first
            }
        }
        return request.remoteAddr
    }
}

